src/common/token_bucket.h

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2018, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file token_bucket.h
+ * \brief Headers for token_bucket.c
+ **/
+
+#ifndef TOR_TOKEN_BUCKET_H
+#define TOR_TOKEN_BUCKET_H
+
+#include "torint.h"
+
+typedef struct token_bucket_t {
+  uint32_t rate;
+  int32_t burst;
+  int32_t read_bucket;
+  int32_t write_bucket;
```

Comment:
Is more of a design questions.

I would love to use that in the DoS subsystem which would mean removing bunch of codes. However, this `token_bucket_t` interfaces kind of bakes the fact that you have "read" and "write" thus a connection.

If that struct had only one bucket, we could use it in some other places in a more generic way. Which would mean we would have `token_bucket_t read_bucket` and the write one in the `or_connection_t`. Is that a major inconvenience? Because as it is, I can use that in DoS code but not really semantically.

Reply:
This is what I was asking you about earlier: if we make the change you suggest here, we have to include the `rate`, `burst`, and `last_refilled_at` fields in the or_connection_t twice.

I can try a refactoring to make this change, but I'd rather do it as a separate branch after this is merged, if that's okay with you.

There are also couple of issues with this implementation that make it a little tricky to reuse. Most notably, it assumes that it's working in units of bytes per second, but it converts them to "bytes per step".  That conversion is fine for bandwidths, but the imprecesion might not be okay for all other units: if so, we'd have to configure it.

Reply:
Right, I think I realized what you meant when you asked when seeing this.

Hmmm... and it makes sense also with our current options since we only have a single bw rate and not a read bw rate and write bw rate.

If we could make this whole piece of work in a generic way to have token bucket for other things than "bytes per second", it would be really great so we don't have multiple bucket implementation. 

I'm not that concerned about having duplicate values that is 4 bytes (rate), 4 bytes (burst) and 4 bytes (timetamp) which means 16 bytes extra. And scaled to 100k connections let say, it is overhead of ~2MB which for a relay is not that much compared to a mobile client.

An option then is to configure a "bucket param" object statically and attach it to any bucket you create with it. The DoS subsystem does not copy rate/burst to the bucket object and uses the global value each time instead. If we had a way to do this with a "param" object, it would also work out pretty well. Which also could reduce by 8 bytes (rate + burst) the `token_bucket_t` object and make it only have a pointer to a static object (+ 4 bytes). Anyway, we aren't saving much space here either way but what I'm more wanting here is a more generic API so it can be used elsewhere. I know that we'll have to go to hidden service circuit creation rate limitation at some point which a bucket system would be _great_ to have.